R/plotAbundanceDensity.R

```diff
@@ -0,0 +1,187 @@
+                                     xlab = abund_values,
+                                     ylab = "Taxa",
+                                     colour_by = colour_by,
+                                     ...)
+        return(plot_out)
+    }
+)
+
+################################ HELP FUNCTIONS ################################
+
+.incorporate_density_data <- function(object, abund_values, n, colour_by){
+    # Gets the assay
+    mat <- assay(object, abund_values)
+    # Gets the most abundant taxa
+    top_taxa <- getTopTaxa(object, n)
+    # Subsets abundance table  by taking taxa of highest abundance
+    mat <- mat[top_taxa, , drop=FALSE]
+    # melt the data
+    density_data <- t(mat) %>%
+        as.data.frame() %>%
+        rownames_to_column("Sample") 
+    # Gets coloring information if 'colour_by' is not NULL
+    if (!is.null(colour_by)) {
+        # Gets information from colData
+        colour_out <- retrieveCellInfo(object, colour_by)
+        # Mirrors back the variable name, if a partial match was used
+        colour_by <- colour_out$name
+        # Adds information to the data frame that includes all density data
+        density_data$colour_by <- colour_out$value
+    }
+    cols <- c("Sample","colour_by")[c("Sample","colour_by") %in% colnames(density_data)]
+    density_data <- density_data %>%
+        pivot_longer(cols = !cols, names_to = "Y", values_to = "X")
+    # Converts taxa to factor. Order of levels is the opposite than in 'top_taxa'
+    # so that taxa with highest abundance is on top
+    density_data$Y <- factor( density_data$Y, rev(top_taxa) )
+    return(list(density_data = density_data, 
+                colour_by = colour_by))
+}
+
+.density_plotter <- function(density_data, 
+                             xlab = NULL,
+                             ylab = NULL,
+                             colour_by = NULL,
+                             layout = "point",
+                             point_alpha = 0.6,
+                             point_shape = 21,
+                             point_size = 2,
+                             text_size = 8,
+                             add_legend = TRUE,
+                             flipped = FALSE,
+                             add_x_text = FALSE){
+    # start plotting
+    plot_out <- ggplot(density_data, aes_string(x="X", y="Y")) +
+        xlab(xlab) +
+        ylab(ylab)
+    
+    # Layout "density" or "point", "density" will be added
+    if (layout == "point"){
+        density_out <- .get_point_args(colour_by,
+                                       shape_by = NULL,
+                                       size_by = NULL,
+                                       alpha = point_alpha,
+                                       shape = point_shape,
+                                       size = point_size,
+                                       position = "identity")
+        plot_out <- plot_out +
+            do.call(geom_point, density_out$args)
+    }
```

Comment:
Still this will produce an empyt plot, if `layout = "whatever"` is set

Reply:
Oh, sorry. My intention was to continue from here and add `else`, that would create a density plot